File: rust/src/safe/play.rs
use std::env;
use std::time::Instant;

use crate::world::World;

pub struct Play;

impl Play {
    const WORLD_WIDTH: u32 = 150;
    const WORLD_HEIGHT: u32 = 40;

    const CLEAR_SCREEN: &str = "\x1b[?2026h\x1b[H\x1b[2J";
    const SHOW_SCREEN: &str = "\x1b[?2026l";

    pub fn run() {
        let mut world = World::new(Self::WORLD_WIDTH, Self::WORLD_HEIGHT);

        let minimal = env::var("MINIMAL").is_ok_and(|value| value == "1");

        if !minimal {
            println!("{}", world.render());
        }

        let mut total_tick: f64 = 0.0;
        let mut lowest_tick: f64 = f64::INFINITY;
        let mut total_render: f64 = 0.0;
        let mut lowest_render: f64 = f64::INFINITY;

        loop {
            let tick_start = Instant::now();
            world.tick();
            let tick_time = tick_start.elapsed().as_nanos() as f64;
            total_tick += tick_time;
            lowest_tick = lowest_tick.min(tick_time);
            let avg_tick = total_tick / world.tick as f64;

            let render_start = Instant::now();
            let rendered = world.render();
            let render_time = render_start.elapsed().as_nanos() as f64;
            total_render += render_time;
            lowest_render = lowest_render.min(render_time);
            let avg_render = total_render / world.tick as f64;

            if !minimal {
                print!("{}", Self::CLEAR_SCREEN);
            }

            println!(
                "#{} - World Tick (L: {:.3}; A: {:.3}) - Rendering (L: {:.3}; A: {:.3})",
                world.tick,
                Self::_f(lowest_tick),
                Self::_f(avg_tick),
                Self::_f(lowest_render),
                Self::_f(avg_render),
            );

            if !minimal {
                println!("{rendered}{}", Self::SHOW_SCREEN);
            }
        }
    }

    fn _f(value: f64) -> f64 {
        // nanoseconds -> milliseconds
        value / 1_000_000.0
    }
}

File: f#/play.fs
open World
open System
open System.Diagnostics

type Play =
  static let worldWidth = 150u
  static let worldHeight = 40u

  static member Run() =
    let world = World(
      width = worldWidth,
      height = worldHeight
    )

    world.PopulateCells()
    world.PrepopulateNeighbours()

    let minimal = Environment.GetEnvironmentVariable("MINIMAL") <> null

    if not minimal then
      printfn "%s" (world.Render())

    let mutable totalTick = 0.0
    let mutable lowestTick = infinity
    let mutable totalRender = 0.0
    let mutable lowestRender = infinity

    while true do
      let tickStart = Stopwatch.GetTimestamp()
      world.DoTick()
      let tickFinish = Stopwatch.GetTimestamp()
      let tickTime = Stopwatch.GetElapsedTime(tickStart, tickFinish).TotalNanoseconds
      totalTick <- totalTick + tickTime
      lowestTick <- min lowestTick tickTime
      let avgTick = totalTick / float world.Tick

      let renderStart = Stopwatch.GetTimestamp()
      let rendered = world.Render()
      let renderFinish = Stopwatch.GetTimestamp()
      let renderTime = Stopwatch.GetElapsedTime(renderStart, renderFinish).TotalNanoseconds
      totalRender <- totalRender + renderTime
      lowestRender <- min lowestRender renderTime
      let avgRender = totalRender / float world.Tick

      if not minimal then
        printf "\u001b[H\u001b[2J"

      printfn "#%i - World Tick (L: %.3f; A: %.3f) - Rendering (L: %.3f; A: %.3f)"
        world.Tick
        (Play._f lowestTick)
        (Play._f avgTick)
        (Play._f lowestRender)
        (Play._f avgRender)

      if not minimal then
        printfn "%s" rendered

  static member private _f(value) =
    // nanoseconds -> milliseconds
    value / 1_000_000.0

Play.Run()
